src/stdlib_math_diff.fypp

```diff
@@ -0,0 +1,120 @@
+#! Inspired by original code (MIT license) written in 2016 by Keurfon Luu (keurfonluu@outlook.com)
+#! https://github.com/keurfonluu/Forlab/blob/master/src/lib/forlab.f90#L2673
+
+#:include "common.fypp"
+#:set RI_KINDS_TYPES = REAL_KINDS_TYPES + INT_KINDS_TYPES
+submodule (stdlib_math) stdlib_math_diff
+
+    implicit none
+
+contains
+
+    #! `diff` computes differences of adjacent elements of an array of the ${t1}$ type.
+    
+    #:for k1, t1 in RI_KINDS_TYPES
+    pure module function diff_1_${k1}$(x, n, prepend, append) result(y)
+        ${t1}$, intent(in) :: x(:)
+        integer, intent(in), optional :: n
+        ${t1}$, intent(in), optional :: prepend(:), append(:)
+        ${t1}$, allocatable :: y(:)
+        integer :: size_prepend, size_append, size_x
+        ${t1}$, allocatable :: work(:)
+        integer :: n_, i
+
+        n_ = optval(n, 1)
+        if (n_ <= 0) then
+            y = x
+            return
+        end if
+        
+        size_prepend = 0
+        size_append = 0
+        if (present(prepend)) size_prepend = size(prepend) 
+        if (present(append)) size_append = size(append)
+        size_x = size(x)
+        
+        if (size_x + size_prepend + size_append <= n_) then
+            allocate(y(0))
+            return
+        end if
+
+        allocate(work(size_x + size_prepend + size_append))
+        if (size_prepend > 0) work(:size_prepend) = prepend
+        work(size_prepend+1:size_prepend+size_x) = x
+        if (size_append > 0) work(size_prepend+size_x+1:) = append
+        
+        do concurrent (i = 1:n_)
```

Comment:
Suggest not making this a `do concurrent`, but rather stick to a regular `do`. The code will not be correct if the loop is executed in parallel. 

Actually the code is correct, because the Fortran standard only requires that `do concurrent` can be executed serially in any order. However this seems to have been a design flaw, and I think most users will assume the parallel interpretation. 

So I suggest using regular `do`.

Reply:
Suddenly my brain is out of order, I will correct it. -_-